api/client/commands.go

```diff
@@ -139,10 +140,38 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 		if _, err := os.Stat(root); err != nil {
 			return err
 		}
-		filename := path.Join(root, "Dockerfile")
+
+		absRoot, err := filepath.Abs(root)
+		if err != nil {
+			return err
+		}
+
+		var filename string
+		origDockerfile := *dockerfileName
+
+		if cmd.IsSet("f") || cmd.IsSet("file") {
+			if filename, err = filepath.Abs(*dockerfileName); err != nil {
+				return err
+			}
+		} else {
+			filename = path.Join(absRoot, *dockerfileName)
+		}
```

Comment:
I meant change the default to the empty string, and use that to determine
when to change it to `<context-root>/Dockerfile`.  Text-parsing for
`CONTEXT/` or `CONTEXT-ROOT/` is going to be flaky IMO.  Then the text can
describe that the default is a file named `Dockerfile` at the root of the
context, but that an argument can be supplied which is either absolute or
relative to the CWD.